File: trees_and_graphs/number_of_islands.js
/**
 * @param {character[][]} grid
 * @returns {number}
 */
var numIslands = function(grid) {
    const valid = (row, col) => {
        return 0 <= row
            && row < m
            && 0 <= col
            && col < n
            && grid[row][col] == "1";
    }

    const dfsRecursive = (row, col) => {
        for (const [drow, dcol] of directions)
        {
            const nextRow = row + drow;
            const nextCol = col + dcol;

            if (valid(nextRow, nextCol) && !seen[nextRow][nextCol])
            {
                seen[nextRow][nextCol] = true;
                dfsRecursive(nextRow, nextCol);
            }
        }
    }

    const dfsIterative = (startRow, startCol) => {
        stack = [[startRow, startCol]]
        while (stack.length)
        {
            const [row, col] = stack.pop();
            for (const [drow, dcol] of directions)
            {
                const nextRow = row + drow;
                const nextCol = col + dcol;

                if (valid(nextRow, nextCol) && !seen[nextRow][nextCol])
                {
                    seen[nextRow][nextCol] = true;
                    stack.push([nextRow, nextCol]);
                }
            }
        }
    }

    const directions = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    const m = grid.length;
    const n = grid[0].length;
    let seen = [];

    for (let i = 0; i < m; i++)
    {
        seen.push(new Array(n).fill(false));
    }

    let ans = 0;
    for (let row = 0; row < m; row++)
    {
        for (let col = 0; col < n; col++)
        {
            if (grid[row][col] == "1" && !seen[row][col])
            {
                ans++;
                seen[row][col] = true;
                dfsRecursive(row, col);
                // dfsIterative(row, col);
            }
        }
    }

    return ans;
};

File: trees_and_graphs/number_of_islands.py
def numIslands(grid: list[list[str]]) -> int:
    def valid(row, col):
        return 0 <= row < m and 0 <= col < n and grid[row][col] == "1"
    
    def dfsRecursive(row, col):
        for drow, dcol in directions:
            next_row, next_col = row + drow, col + dcol
            if valid(next_row, next_col) and (next_row, next_col) not in seen:
                seen.add((next_row, next_col))
                dfsRecursive(next_row, next_col)
    
    def dfsIterative(start_row, start_col):
        stack = [(start_row, start_col)]
        while stack:
            row, col = stack.pop()
            for drow, dcol in directions:
                next_row, next_col = row + drow, col + dcol
                if valid(next_row, next_col) and (next_row, next_col) not in seen:
                    seen.add((next_row, next_col))
                    stack.append((next_row, next_col))
    
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    seen = set()
    ans = 0
    m = len(grid)
    n = len(grid[0])

    for row in range(m):
        for col in range(n):
            if grid[row][col] == "1" and (row, col) not in seen:
                ans += 1
                seen.add((row, col))
                dfsRecursive(row, col)
                # dfsIterative(row, col)
    
    return ans
